### ISSUE TITLE
Unexpected time-specific values in topics_over_time (an indexing issue?)

### ISSUE THREAD

**Issue by sokol11** [MaartenGr/BERTopic#128]
Hi. When I run `topics_over_time`, I noticed that time-specific values (words) completely mismatch the global topic words in some instances. In fact, they appear to be injected from another topic altogether. This behavior is very odd, as it seems to happen for some topics only.

My leading hypothesis is that it has something to do with the fact that I ran `.topics_over_time` before I updated the model with a different `n_gram` setting by running `.update_topics()`.

My second hypothesis is that I excluded all docs with the `-1` label when I ran `.topics_over_time()` but when I similarly tried to run `.update_topics()` with `-1` records excluded, it returned an IndexError after some time.  `.update_topics()` ran fine when I included all records originally transformed by the model. So perhaps excluding some records at certain analysis steps messes with the model index?

Is there any guidance you can provide on whether excluding certain docs (particularly the ones with the `-1` label) may introduce errors in the downstream analysis? Also, does excluding these records actually cut down computation time, or do they already get ignored by default? Last, does `.update_topics()` invalidate some of the other analysis tasks, like `topics_over_time`, so they have to be re-run after `.update_topics()`?

Thanks a million!

**Comment by sokol11** [comment]
So after I re-ran `.topics_over_time()` on the entire data set, the problem went away. So it must have been one of those two things: either the fact that I updated the model, or that I excluded the `-1`-labeled docs when I first ran it.

**Comment by MaartenGr** [comment]
Glad to hear that the problem went away. In your first post, you were definitely spotting what could happen. 

The first thing you mentioned depends on what you expected after running `.update_topics()`. This method does not change the topics over time you created before with `.topics_over_time()`. If you simply change the order and run `.update_topics()` first, then you should see the new topics in `.topics_over_time()` assuming you also passed the new topics to `.topics_over_time()`:

This should be the correct order:

```python
# Create model
topic_model = BERTopic()
topics, _ = topic_model.fit_transform(docs)

# Update model
topic_model.update_topics(docs, topics, n_gram_range=(1, 3))

# Apply DTM
topics_over_time = topic_model.topics_over_time(docs, topics, timestamps)
```

Second, and I think this mostly affected the representations, is that excluding topics with the -1 can mess with the indexing especially depending on the way you excluded those. Performance-wise it does not really matter whether you remove a topic. The c-TF-IDF calculations are all matrix calculations and therefore are still performant even if you remove a certain amount of topics. The number of bins or unique timestamps is what can slow down computation time significantly. 

Third, it is important to realize how BERTopic extracts the topic representations at each time. It does so by re-calculating the c-TF-IDF scores for only the documents that are in that specific timestep. This also means that if you only have a few documents, the topic representation will not be as accurate as when you would have many documents. Lowering `nr_bins` in `topics_over_time()` could help you in creating better topic representations as larger clusters of documents are created. Personally, I would advise setting `nr_bins` to mostly 10 and only use higher values if you have more than enough data to create good representations at 10.

**Comment by sokol11** [comment]
Thank you, that all makes perfect sense.

As a quick follow-up question:

I would like to compare `topics_over_time` between classes. Originally, I thought of computing `topics_over_time` separately on a subset of documents (class 1, class 2, etc.) and comparing the figures. But now, I am not sure since you say that computing `topics_over_time` on a subset is less precise? Is there an alternative way?

**Comment by MaartenGr** [comment]
Hmmm, if you have enough data available, I would indeed compute `topics_over_time` separately for each class based on only their documents. However, this does depend on the amount of data that you have available. You are already slicing the number of documents by spreading them out over different timesteps. Further slicing them by classes could reduce the number of documents that topic representations make little sense. For example, extracting the topics from only 2 documents is not really accurate. Thus, make sure you have enough documents to do so.
